09_json/n0npax/cmd/puppy-server/main.go

```diff
@@ -0,0 +1,94 @@
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"io"
+	"io/ioutil"
+	"log"
+	"os"
+
+	"gopkg.in/alecthomas/kingpin.v2"
+
+	puppy "github.com/anz-bank/go-course/09_json/n0npax/pkg/puppy"
+	store "github.com/anz-bank/go-course/09_json/n0npax/pkg/puppy/store"
+)
+
+var (
+	logFatalf = log.Fatal
+	parser    = parseArgs
+)
+
+func main() {
+	config, err := parser(os.Args[1:])
+	if err != nil {
+		logFatalf(err)
+	}
+	logFatalf(runPuppyServer(&config))
+}
+
+func runPuppyServer(c *config) error {
+	if err := createStorer(c); err != nil {
+		return err
+	}
+	if err := feedStorer(*c); err != nil {
+		return err
+	}
+	return nil
+}
+
+func parseArgs(args []string) (config, error) {
+	var storeType string
+	var puppyFile *os.File
+	kingpin.Flag("data", "path to file with puppies data").Short('d').FileVar(&puppyFile)
+	kingpin.Flag("store", "Store type").Short('s').Default("map").EnumVar(&storeType, "map", "sync")
+	_, err := kingpin.CommandLine.Parse(args)
+	return config{puppyFile, storeType, nil}, err
+}
+
+func createStorer(c *config) error {
+	switch c.sType {
+	case "sync":
+		c.storer = store.NewSyncStore()
```

Comment:
I think it would be nicer if this function didn't modify it's input but return the Storer:
```go
func createStorer(storeStr string) (puppy.Storer, error) { //...
```
To me config should as much as possible only hold read-only/immutable config info (which storer is not).
Also, where all relevant information is just one (or two) fields of the config I find it nicer to pass those few fields (keep the interface smaller).

Reply:
fixed in `lab10`